### DISCUSSION TITLE
[API Discussion] Existing single-valued fields that should be multi-valued

### DISCUSSION THREAD

**Discussion by dweymouth** [opensubsonic/open-subsonic-api#8] [API requests]
### Type of change

API extension

### Proposal description

I think it makes sense to cover all of these as a single discussion, since the way we'd handle it should be the same for all the fields, and different from handling new metadata fields like artist roles, etc.

The following are existing Subsonic API metadata fields that should be multi-valued:
* Artist (both albums and tracks should be able to have multiple artists)
* Genre (both albums and tracks should be able to have multiple genres)
* any others anyone can think of please add in the comments

### Backward compatibility impact

We need to discuss if there's a way to do this without backward compatibility impact. A potential way could be to just add Artists and Genres as new fields to the album and song payloads, but this is a little awkward as the Artist and Genre fields would still exist and only have one of the potentially multiple entries from the Artists and Genres.

### Backward compatibility

- [X] No backward compatibility impact.

### API details

Context above. We need to discuss what this would actually look like, API-wise.

**Comment by Tolriq** [comment]
I had not yet the time to port the rest of the proposals but there's a lot to see about extending the metadata with new fields and everything.

Since we should differentiate album artists from song artists it should be all talked at the same time as it's connected to the multi values. Same for artist roles like composer and everything.
Need to ensure all can be covered without impacts on the namings / structure.

**Comment by gravelld** [comment]
> We need to discuss if there's a way to do this without backward compatibility impact. A potential way could be to just add Artists and Genres as new fields to the album and song payloads, but this is a little awkward as the Artist and Genre fields would still exist and only have one of the potentially multiple entries from the Artists and Genres.

How about only supporting this in an extension? It would be a little confusing for client authors, but this is only something that needs to be understood once, by the programmer of the client.

**Reply by Tolriq** [reply]
That would require new API endpoints for nearly everything that returns data, in that case it's better to build a new API from the start, something that was already ruled out.

**Comment by lachlan-00** [comment]
The way I'll be doing this in Ampache is to add the plural element for each field. Keeping the original single values as the main/primary.

e.g. a song/album has a primary artist and then additional artists attached to the object

so the response returns
* artist:
* artists: []
* albumartist:
* albumartists: []

This will let me keep things back compatible and can be ignored by clients.

**Comment by sentriz** [comment]
since beets.io has [nearly added support](https://github.com/beetbox/beets/pull/4743) for this stuff it has peeked my interest

@dweymouth @lachlan-00 you both mentioned how to tackle backwards compatibility and i like the idea, just thought i would make sure we're on the same page:

## problem

consider this [musicbrainz release group](https://musicbrainz.org/release-group/95469885-a4de-4192-b7b7-d13d9a445cf1), the tags of a track on this album may look like

```
albumartist="Marc Hurtado & Alan Vega"
albumartists=["Marc Hurtado", "Alan Vega"]
```

with the current state of subsonic clients and servers we just have a single album artist named `"Marc Hurtado & Alan Vega"`, completely independent from any other artists named `"Marc Hurtado"` or `"Alan Vega"` that the user might have

#### getArtists
|id|name|
|---|---|
|ar-1|Marc Hurtado & Alan Vega|
|ar-2|Marc Hurtado|
|ar-3|Alan Vega|

#### getAlbums2
|id|name|artist|
|---|---|---|
|al-1|Sniper|ar-1|

here have no connection to ar-2 or ar-3, which is not so nice for the user :-1:

## proposal

considering the same release, we only store those individual artists 

#### getArtists
|id|name|
|---|---|
|ar-1|Marc Hurtado|
|ar-2|Alan Vega|

#### getAlbums2
|id|name|albumartist|albumartists|
|---|---|---|---|
|al-1|Sniper|ar-1|[ar-1, ar-2]|

so just an xml/json array like thing as already suggested

now the user sees a much cleaner view of their album artists, and from a single album page, can click on either of the 2 albums artists to check them out :+1:

-------------


does this line up with what was suggested already?

and how about for clients that don't support the extension? from their perspective the album artist is just "Marc Hurtado" and there's no mention of Alan Vega anywhere. maybe that's fine if was the Marc Hurtado was the primary artist, and he was the first in the list in the audio metadata. but maybe that's not always the case. is there a better way to do backwards compatibility?

**Reply by Tolriq** [reply]
The real question is not about the displayed text but how the servers currently handle the artists and ids themselves.

In all cases without even talking about returning new values the servers that want to support both cases needs to handle this in a special way   to keep generating fake artists with fake ids.

So once the server handle both the legacy mode internally and the new mode this is not really an issue, you return the legacy data in the `artist` and `artistId` and the new data in `artists` and `artistIds` the API exposure is not the problem, it's more how the servers want to address both modes internally first.

**Reply by sentriz** [reply]
ok, but i'd just like to have some concrete of examples of expected behavior before talking about internals.

eg, in this example:

> legacy data in the `artist` and `artistId` and the new data in `artists` and `artistIds`

in the case of the album [Sniper](https://musicbrainz.org/release-group/95469885-a4de-4192-b7b7-d13d9a445cf1) that i mentioned, what do the legacy `artist` and `artistId` values look like? and what do they point to? 

would it be to the original raw album artist like "Marc Hurtado & Alan Vega", or just the first artist like "Marc Hurtado" as was already suggested

**Reply by Tolriq** [reply]
That's what servers currently do. 
This is not really documented, most have an artist having both names in it and an ID for it.
Some may only return the first artist.
Those should not change since it's legacy and it's the currently returned value. Only the new values should be changed. Else this is not related to the API change but to internal server changes and the API should not really have a word on that.

But this is tied to internals of servers, I don't know if there's already servers that support multiple artists internally and yes expose consolidated single one to the Subsonic API.

**Reply by sentriz** [reply]
considering a server that supports both legacy and improved artist modes, how should the server decide on what to return for a request to getArtists? it needs to decide between the two modes

**Reply by Tolriq** [reply]
The new data is in new fields.

How servers handle the legacy data is internal to each servers and their decision.

This is not related to the API but to how they handle the migration internally.

**Reply by dweymouth** [reply]
My first thought is that it should be treated how multi-genres are currently treated. So if you have an album with two album artists, then GetArtists should include an artist entity for each, and that album should appear under both. But then the question is what should it return in the single artist field for the album? In the case of genres, most servers return just the first genre. In the case of artists, you'd expect it to return both artists names (eg "Marc Hurtado & Alan Vega") but then that joint artist wouldn't have an artistID under the new system. So I really don't know what the best thing to do is. Maybe returning "Marc Hurtado & Alan Vega" as the single artist name, but having the singular artistID for the album just point to Marc Hurtado?

I guess it could be unspecified what the server returns for the singular artist and artistID fields  - so the server could choose to return only the first one, or to concatenate them in the artist name string but return the singlar artistID for just the first one. What I *don't* think should be done is to create an artistID for every set of artists that appear together in an album. At least don't return that "multi-artist" in the GetArtists list, as that would break clients that support multi artists properly.

**Reply by sentriz** [reply]
> How servers handle the legacy data is internal to each servers and their decision.

I think it is, users have data in both their albumartist tag, and their albumartists tag. the server needs to decide which to return for the getArtists call no?

**Reply by sentriz** [reply]
@dweymouth I agree I had similar thoughts above 👍

**Reply by Tolriq** [reply]
The thing really is that the current behavior is at server choice. The files already have the multiple tags and the servers already have made a choice at what they do. The first, concatenate, ...

OpenSubsonic is  currently not about enforcing breaking changes on the current API so we have nothing to enforce on the current fields as the behavior is not documented.

We can only force the data on the new fields.

**Reply by dweymouth** [reply]
I think the point that @sentriz is making is that the servers may make a different choice on what to do with multiple (album-)artist tags now that multiple values will be supported. For example a server might be choosing to concatenate right now, and just create an artistID for the concatenated multi-artist, but now they will treat them as separate entities as described above. Since it's undocumented currently what the server will do with multiple tags, it's not strictly a breaking change but it would be a change in behavior. I also think it's fine, since arguably multiple artist tag handling was "broken" with the current server behaviors

**Reply by lachlan-00** [reply]
The Ampache multi artist setup is to rely on MBID to separate groups of artists using the Artists tag and the mbid tags.

I can go into more detail but from a client perspective the client shouldn't care how the server gets it's artists.

Currently the Subsonic implementation in Ampache uses the Primary/First artist in results. so it's not "Marc Hurtado & Alan Vega" it would be one of those artists. (split on tags by the server)

In the ampache web/api you would get both artists as proposed in the parent comment (https://github.com/opensubsonic/open-subsonic-api/discussions/8#discussioncomment-5661434) ar-1, ar-2

I don't think you need to get too technical here but the reason i use the plurals in the api is it can be safely ignored and doesn't affect the single value data.

Because i've already implemented it it's just a matter of me adding extra data to the response. for subsonic implementations it seems like there is a lot of structural work in the backend/database to be done?

**Reply by epoupon** [reply]
Indeed, the goal open subsonic is also to clarify docs for existing endpoints. But as Tolriq said, I am afraid there are too many different implementations. Maybe we can use the "should" word in the doc to try make things better, but the new api is the right answer anyway.
(LMS concats the names and sets artistId only if there is a single artist)

**Reply by Tolriq** [reply]
Since OpenSubsonic is adding new fields that will be consistent in all cases as supporting both single and multiple artists, I do not think there's any benefits on implying that OpenSubsonic servers should update how they handle the older data.
Either they keep doing what they do and do not support the new fields and we should not have anything to say.
Or they support the new fields and then what is present is the older fields is not really related to Open Subsonic.

Anything we enforce can make things harder for the servers to manage their transition in handling both fields if they need internal rewrite.
So let's keep things simple and limited to the purpose of this project.

**Reply by lachlan-00** [reply]
We've gotten a bit long winded, should probably update the original proposal/comments now?

I suggest that this is limited to adding (for this change and all optional additions)

* song objects
  * add artists
  * add artistIds
  * add genres
* album objects
  * add artists
  * add artistIds
  * add genres
* artist objects
  * add genres

**Reply by Tolriq** [reply]
I would have merged all the new "data" fields addition in a single update (So also mood, bpm, ...).
Since all the fields are optional I'm not sure there's a need to handle each cases differently no?

We make the list of new supported fields and their definition. All the fields are optional. This is one of the simple non breaking change.

**Reply by lachlan-00** [reply]
we can do that too. i was just thinking that some servers may not be able to support all new fields in one go so doing them in groups would let servers say "supports multi-artist", supports x

**Reply by Tolriq** [reply]
No server will add them in one go, but it does not matter, they add the field as they have them. Some servers already have some in db and it's just exposing them. Some will need to scan the data, ...
Everything is optional data.

The only difference with current API is that when a field is supported by a server it must be returned with empty data when no data, so that clients can know that the data is missing and it's not the server not supporting the field.

**Comment by sentriz** [comment]
taking the subsonic "album" as an example it has the fields "artist", and "artistId". extending this, the natual thing seems to add 2 extra. giving us a total of "artist", "artistId", "artists", "artistIds"

for example

```
"album": {
  "id": "al-6",

  "artistId": "ar-1",
  "artist": "Alan Vega",

  "artistIds": [ "ar-1", "ar-2", "ar-3" ],
  "artists": [ "Alan Vega", "Alex Chilton", "Ben Vaughn" ],
}
```

though it feels a bit clunky, we assume that artistIds and artists are the same length, and sorted the same way. but it leaves room for mistakes

---

there's also the option of a single array with a more complex object like

```
"album": {
  "id": "al-6",

  "artistId": "ar-1",
  "artist": "Alan Vega",

  "artists": [ 
    {"id": "ar-1", "name": "Alan Vega"},
    {"id": "ar-2", "name": "Alex Chilton"},
    {"id": "ar-3", "name": "Ben Vaughn"}
  ],
}
```

though i don't know how that would translate to XML, or if we have anything like that elsewhere in the spec

thoughts on these options?

**Reply by Tolriq** [reply]
Both works and both style are used by other media centers. The first is more consistent with current API, the second is generally better but might be harder for XML and maybe some servers / clients.

I do not have strong preferences, but if we move to the second choice, then it became the new way that we should apply for the future changes. Consistency in the API is important.

**Reply by sentriz** [reply]
@lachlan-00 @epoupon - i believe ampache and LMS support multiartist 😎 do they expose this over the subsonic API? if so, how? (in terms of field names, structure, etc)

**Reply by epoupon** [reply]
Indeed LMS supports multiartists (also with "roles" like, multi albumartists, multi composers, multi performers (with performance roles), etc.)
As the current API does not support multiple artists, the artists are string-concatenated and exposed as a single one, and the artistId field is then omitted... Really very clunky, but I found this better than just picking the first one of the list...
Edit: and for the webUI, as it is mostly server side rendered, info is gathered using a C++ lib so no help to decide a json structure (even if it is more like the second option)

**Reply by Tolriq** [reply]
So @opensubsonic/servers @opensubsonic/clients I'll soon start to work again on finalizing the proposals to start moving on now that the big holidays are over.

We previously assumed to keep things similar to current API for some of the new things but @sentriz opened the pandora box about making more improvements to the multi fields by returning objects and not having to return 2 different fields that needs to be in the same order. (https://github.com/opensubsonic/open-subsonic-api/discussions/8#discussioncomment-6908113)

Since this is globally better in the end, if you have objections about it, please expose the reasons. If you have experience with the XML part please also give feedback.

**Reply by dweymouth** [reply]
This is how the XML would have to work (example for getAlbumList2) :

Current:
```xml
<albumList2>
  <album id="1768" name="Album 1" coverArt="al-1768" songCount="2" created="2002-11-09T15:44:40" duration="514" artist="A Single Artist" artistId="829"/>
  <album id="2277" name="Album 2" coverArt="al-2277" songCount="4" created="2004-11-28T00:06:52" duration="1110" artist="Double Artist 1, Double Artist 2"/>
  ...
</albumList2>
```

New:
```xml
<albumList2>
  <album id="1768" name="Album 1" coverArt="al-1768" songCount="2" created="2002-11-09T15:44:40" duration="514" artist="A Single Artist" artistId="829">
    <artist name="A Single Artist" id="829"/>
  </album>
  <album id="2277" name="Album 2" coverArt="al-2277" songCount="4" created="2004-11-28T00:06:52" duration="1110" artist="Double Artist 1, Double Artist 2">
    <artist name="Double Artist 1" id="123"/>
    <artist name="Double Artist 2" id="456"/>
  </album>
  ...
</albumList2>
```

**Reply by dweymouth** [reply]
Even though this is verbose I think it's better for XML too than the alternative, since there is no list type in XML and for listing out the multiple artist names and IDs, we'd need to choose a separator character, like `;`, and have a way to escape it if it appeared in an actual artist name.

like
```xml
<album name="foo" artist="first artist" artistId="1" artists="first artist; second artist" artistIds="1; 2"/>
```

**Reply by Tolriq** [reply]
Ok so we do not define a field name, I assume we could also add a bunch of `<genre />` too? 
But we still need to define some new subTypes for those to ensure that if we add some required field to the artist type they are not mandatory here too by leak no?

Edit: For your last post, so i guess this is the way to go then :)

**Reply by dweymouth** [reply]
See the XML schema here: http://www.subsonic.org/pages/inc/api/schema/subsonic-rest-api-1.16.1.xsd

If you search `AlbumWithSongsID3` you can see how it specifies the nested song objects with `<xs:sequence>`. We'd want to add an `<xs:sequence>` for artists to the base type `AlbumID3`. Actually now that I think about it I'm not sure if we'd want to wrap it even more like:

```xml
<album id="al-1" name="foo" ...>
  <artists>
    <artist id="ar-1" name="bar"/>
    <artist id="ar-2" name="baz"/>
  </artists>
  <genres>
    <genre name="Rock"/>
    <genre name="Pop"/>
  </genres>
  <song name="track 1" .../>
  <song name="track 2" .../>
</album>

**Reply by dweymouth** [reply]
Currently the songs list is not wrapped within another element tag, it's just directly nested under <album> but if we are now going to have a sequence of artists and maybe genres as well, it may make sense to wrap those sequences under new <artists> and <genres> tags so they're separated from the song list. Of course we can't wrap the song list in a <songs> tag for backwards compatibility.

**Reply by sentriz** [reply]
what is the benefit of the wrapping again? to test it, i updated gonic (go code, naively just putting a list of artist objects in the album)
![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/9de94e92-1416-4aea-be35-1685319b441b)


and it produces this, which seems usable right? it's the name technique as the song list in the same object


![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/c0c648e7-f926-4e16-ab25-cc84711a8d81)
![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/2c3d0561-17d6-4213-a6bf-f4fce62807b5)

**Reply by sentriz** [reply]
search3 also works this way IIRC

**Reply by Tolriq** [reply]
I personally have 0 opinion about the best XML way.

**Reply by dweymouth** [reply]
I think that works, I was just thinking about avoiding the possibility of:

```xml
<album>
  <artist/>
  <song/>
  <genre/>
  <song/>
  <artist/>
  <song/>
</album>
```

but any half-decent XML parser should be able to handle this fine anyway

**Reply by sentriz** [reply]
OK should we roll with that then? I think not having the nesting will make things a little cleaner for clients too

**Reply by Tolriq** [reply]
For the record currently the doc does not cover XML, I hope to get some help to write that part, as it was already a tons of work to build the current doc, and it will take time to prepare the JSON proposals.

**Comment by epoupon** [comment]
Naïve question: in some of your examples in this whole discussion, we sometimes can read "albumartists", and sometimes "artists".
But do we agree we want to just use "artists" and the parent element is used to know the context? (in "album" => "album artist", in "track" => "track artist", like it is in the original Subsonic API).
And in getArtists, we list them all.

**Reply by Tolriq** [reply]
Well IMO no they are different and should be exposed differently properly.
1) To ensure that the returned value correspond to what we expect. Artists on album can be the album artists or the list of all artists of the albums or both.
2) getArtists should be able to return both only albumartist and all artists with a flags, as many people do not want to see hundreds of track collaborators. (Even worse if we expose composers, lyricist, ... as artists to get their metadata).
3) Being able to easily display the different data on the right place without needing to do additional queries.
4) Being able to easily filter albums where artists are album artists and albums where they are collaborators. So should be different filters.

Those are frequent needs from users, and all other media providers properly expose both data. (I personally have a full copy of the database and recreate the data from the tracks and albums, but this would help faster syncs and would help all the online first apps to propose the same features).

**Reply by epoupon** [reply]
Ok, but then for artists in general this is more than just duplicate existing "artist" fields to handle multiple values. Maybe this would deserve a separate discussion?

**Reply by Tolriq** [reply]
Not sure to follow what part you are referring to?

**Reply by epoupon** [reply]
I just mean that handling multi values for artists is more complicated than for genres if we want to refactor things like you proposed.

**Reply by Tolriq** [reply]
Not that much, but everything is split in multiple thread already, with the support of composers and everything. I'll try to build a full proposal for the field extensions soon.

**Comment by tamland** [comment]
Not sure if this has been mentioned, but many releases use specific order and words for multiple artists.

Examples:
https://www.discogs.com/master/143760-Alpha-Omega-Meets-Dub-Judah-Almighty-Jah
https://www.discogs.com/release/1663520-Alpha-And-Omega-Featuring-Jonah-Dan-Paul-Fox-12-And-Dan-I-Songs-From-The-Holy-Mountain
https://www.discogs.com/master/1967440-Jonah-Dan-Alpha-Omega-Spirit-Of-The-Ancients

Any ideas how this can be visually preserved by clients? I think the spec should say something about order at least (such as the server should return them as they appears)

**Reply by sentriz** [reply]
gonic will try its best to preserve the order that they appear in the tags, but the delimiter will be lost like " and " or " & "

(in the PR to airsonic-refix I just made it use a comma as you may have seen)

**Reply by sentriz** [reply]
the way musicbrainz solves this btw is by keeping track of a "join phrase" per mentioned artist 

(that said, i don't think anyone puts that stuff in their id3/vorbis data, or even if there's a standard for it)

![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/627d839d-4aec-4b1a-9494-4b5ea67bf609)
![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/32916fa1-0a15-46f9-9685-a36e749841a5)

**Reply by Tolriq** [reply]
1) Order
This is a pain point for most providers :( They most of the time return the data sort alphabetically. 
While this is a nice to have and clients should try to keep the order that provider provide, I'm not sure that we should enforce this as it can be hard for servers. Maybe just a note about should try to keep tag order ?

2) Join / display value
About the joints this is a point some providers support by returning a displayArtist and displayAlbumArtist fields so that clients can use that precise value and do not try to recreate the display value themselves. I'm not sure I already talked about those field somewhere here. But if servers can / want to support those I'd gladly expose them to the API.

**Reply by sentriz** [reply]
> Maybe just a note about should try to keep tag order ?

i agree on the note :+1: 

> some providers support by returning a displayArtist and displayAlbumArtist fields so that clients can use that precise value and do not try to recreate the display value themselves 

i'm not sure that's entirely useful here, the client could display it, but it still needs to generate clickable links for individual components of it. eg `[A](/artist/A), [B](/artist/B), and [C](/artist/C)`

**Reply by Tolriq** [reply]
Well on Symfonium I open a bottomsheet with the actual list of artists if there's more than one. Having items directly clickable on a phone screen is hard to click + need to scroll if a few long names + ...

So it's both world, the proper display value as intended and link to all the artists. Most clients can also easily match the names in the display value and create the clickable entries if they want.

I for sure use the value when available and users like it (Specially because it workaround the issue with order of the tags, that value is always the proper one).

**Reply by tamland** [reply]
The musicbrainz solution seems nice and would work as a possible extension to the proposal at least. (how to actually obtain this data is a different story though). Same with "Artist as credited".

**Reply by Tolriq** [reply]
As a reminder there's also the sort artist, sort album artist composer sort to handle. IMO the MB way is only useful to generate the displayXXX values. I don't see any provider storing that low level data to regenerate a display value each time. That's a lot more complex to handle at a database level. Specially for all those who do not even handle the order of the artists.

**Comment by sentriz** [comment]
following the decision made on getAlbum's `artists` field, are we okay with following the same pattern for `genres`? (but just an array, we don't need an object this time)

for example:

JSON:

![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/e68389cd-b9c1-4935-9f31-53378752f91f)

XML:

![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/0c343079-783e-4474-8300-eaeb734a1319)

**Reply by Tolriq** [reply]
You are going faster than the music, nothing have been acted yet unfortunately. We need to write the full proposal and vote on it.
Artists needs to take in account album / song artists differentiation and the API should be ready to handle the other contributors.

But yes since subsonic Genres do not have ids it should be a simple list of strings.
No reactions in months and now everything should be done in 2 days :)

**Reply by sentriz** [reply]
> We need to write the full proposal and vote on it

i agree, and i believe @dweymouth and @tamland are already okay the format of `{"artists": [ {"id": "a", "name": "b"}, {"id": "x", "name": "y"}] }`

so that should be already be at least 3 votes for that style.

can i help writing the proposal somehow?

> No reactions in months and now everything should be done in 2 days :) 

i'm not expecting anything, just a mini proposal on this discussion to see what people think

> Artists needs to take in account album / song artists differentiation

i don't think this is relevant for the getAlbum response right? i can't imagine we would ever consider track-artists there

**Reply by Tolriq** [reply]
You already have merged the code :p 

And yes it is relevant to the getAlbums that's the whole point of albumartists and artists. That's how you differentiate albums from compilations. And how we name things to ensure consistency and coherence in whole API. Handling of compilation is not possible currently with Subsonic, the artists rework needs to allow to properly support that and differentiate things.

And in my case it also matters that everything is properly linked everywhere so I can properly fast sync. (So every changes must be available from the search3 endpoint too, until we have a new API to allow fast sync of the whole database)

You should be able to filter albums where artist X is album artist but also filter all albums where artist Y appears either on a track only or as composer if you wish.

It does not make sense to have to query the songs to get the albumid then ask for the album details.

The proposal and implementation needs to take in account everything that was posted in all the threads here, so not do something that will block a feature we already know we might want.

So for the artists in total : Handling of album artists /artists, compilation identification, other roles (composer at least, and others so should be role agnostic, but with well know values, ....) And I probably forget some details, I need to take time to re read everything posted, be sure to not have missed stuff that other more robust providers offers, then finalise something that works for everything but still be open enough to allow partial support of each things. [Like proper values to return to differentiate no value available from server vs server does not support that field at all, ...]

**Reply by sentriz** [reply]
okay i will let you guys figure out track artists then. personally i have no use for them, gonic doesn't even support them. (an artist is always an album artist, track artists are display only)

**Reply by Tolriq** [reply]
So you do not support compilations at all? That's a big pain point for users on Subsonic. If you have a compilation of a top 100 rock songs. You have 100 artists for your album instead of a Various Artist one and a compilation flag. Or all the songs have Various Artist and it's also unusable.
Same for many rock / pop albums with guests. An album X made by Artist Y have the album Artist Y. If on multiple songs he collaborate with A, B, C and Z then they are not artists of the album and by default should not be seen as such.
Supporting multiple artists without properly supporting the difference between those will just make your database filled with hundreds of artists that will cripple the user experience, most users only care about album artists by default for albums, but still want to know and filter the songs with the actual artists on the songs.

What tags do you extract?

**Reply by sentriz** [reply]
if the music is tagged correctly, then the tracks in the compilation's albumartist field will be set to "Various Artists". or it could be anything, like a single artist's name (usually the main artist of the album, not contributors)

so long as it's set to the same thing across all the tracks of the compilation, the artists list never gets clutter 

---

and specifically to gonic, it only checks the albumartist field of the first track in the album, just in-case someone accidentally put different albumartists across the tracks

**Reply by Tolriq** [reply]
But the tracks do have artists. So you have albums with Artist A but all songs are from different artists, and tracks do not expose artist A.
They are just magically linked.
And depending on the need you either request albums from artist or all songs by artists then 1 query per album to list the albums where an artist collaborate.

So the global artist list will either be lacking all the artists that are not album artists so songs will be without artists.
Or the global artist list will be cluttered with all the artists without anyway to differentiate songs artist from album artists.

There's something wrong in both case. But from your last comment it seems that it's case A where tons of artists information is not available and usable.

**Reply by sentriz** [reply]
i see what you mean that it could be useful in some cases to have a list of track artists instead of album artists, but maybe that's a separate proposal to this one.
right now getArtists nearly always returns albumartists right? are there any subsonic servers right now that would return track artists there? seems crazy to me. even in the [example response](https://opensubsonic.netlify.app/docs/endpoints/getartists/) we mention "albumCount" which implies album artists only surely. 

when talking about the subsonic "album" object, this proposal is only concerned with field X that should now be field [X], and I would guess that, for existing subsonic servers, X is based on the albumartist field and should now be based on the albumartists field instead

**Reply by Tolriq** [reply]
Some servers returns track artists the album count is just 0 and that make sense and help differentiating them. But an artist can be album artist on an album and guest and another.

And I'm not even starting to talk about Classical handling.

And this proposal is not about album only, it the global way of handling multi values for everything.
And artists is just a part of that as composers and everything.

The API needs to be consistent and handle everything correctly.

But for my case of Symfonium and offline fist app.
I need a way to get all artist to sync them.

If getartists returns not all artists but songs returns IDs that are not present in that list what does it give?

I need to do one request to get artist for each unknown IDs in hope that it returns something? 

Highly inefficient and broken API.

Same for online first app.  How can an app display all artists to get all their songs. 
For many users it's not all just about albums.

And again Classical is an even more complicated beat.

**Reply by sentriz** [reply]
for gonic, since track artists are not supported (for now) the track's artistId is the album's album artist. so consistent with getArtists.

but still, my understanding is we're updating field X to be [X]. we shouldn't change the meaning of X. if it was albumartist, it should now be albumartists. if it was artist it should now be artists.

picard and beets can set these multi value tags

i suppose its up to different servers but what makes sense to me, in general is

album.artists is album artists
song.artists is track artists 

in gonic these happen to be the same, but still I'm not proposing to change the meaning of the fields we have already. that could be a separate proposal to define that better

**Reply by epoupon** [reply]
> are there any subsonic servers right now that would return track artists there? seems crazy to me. even in the [example response](https://opensubsonic.netlify.app/docs/endpoints/getartists/) we mention "albumCount" which implies album artists only surely.

LMS returns them all by default (both track and album artists), and there is a per-user option to return only "album artists" in the search endpoint...
"albumCount" is never set to 0 for track artists, it is the count of albums where the artist appears on (very handy for compilations).

In LMS, an artist is just an artist whatever the tag that has been used to extract it (like artist, albumartist, composer, remixer, conductor,  performer, etc.). A track or an album is not linked to an artist, but to an artist/role entry.

Ideally we should be able to filter an artist for a given role:
* I want to see all the albums composed by Mozart.
* I want to see all albums or even just tracks sung by Bruce Dickinson 
* etc.

**Reply by sentriz** [reply]
:+1: i agree it would be nice to have better filters like that. but in terms of album.artist becoming album.artists (or track.artist becoming track.artists), what would that mean for LMS? given that the user can configure how they like their artists to be set up

**Reply by Tolriq** [reply]
What LMS does is how all most providers do and how it should be handled.

By having song track artists being in fact album artists, you hide very valuable information. That's a blocking point for me for example.

And we must be able to return all artists by a way or another.

So the simple fact that already 2 servers have different understanding of the values, confirm that there's the need to clarify things and your sentence: 

```
i suppose its up to different servers but what makes sense to me, in general is

album.artists is album artists
song.artists is track artists

```

An API value can't be up to the server, that's what currently make Subsonic API a complete mess and one of the reason of this project. Having all servers returning the same data with the same meaning from the same endpoints.

So while I agree your definition can make sense, it's currently not defined and is part of the proposal. We can totally agree on your definition without preventing a filter on getAlbums with contributors that will not be part of the returned data.

But we **need** to be clear on what is what to ensure that we can support roles and track artists.

On a personal note I want to ditch Jellyfin for a Subsonic for my large collection but need multi artists (track and album) and multi genre and hope for Composer to have the same navigation abilities.

**Reply by Tolriq** [reply]
Ok so @opensubsonic/servers @opensubsonic/clients I've planned to do the draft on multiples values and most common missing data from all the current issue next week. If you have some ideas you did not wrote, you can just add some details in this thread so I compile them in the global draft.

**Comment by Tolriq** [comment]
See proposal https://github.com/opensubsonic/open-subsonic-api/discussions/52
